### PR TITLE
added algo for finding permutations of an array

### PR DESCRIPTION
**Pull request by rajansh87**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: data_structures/arrays/permutations.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# All permutations:
+def permute(nums: list[int]) -> list[list[int]]:
```
Comment (cclauss) [inline data_structures/arrays/permutations.py:2]:
```suggestion
def permute(nums: list[int]) -> list[list[int]]:
    """
    >>> from itertools import permutations
    >>> numbers in ([], [1], [1, 2], [2, 1], [-1, -3, -2], [5.5, 3.3, 4.4, 1.1, 2.2])
    >>> all(permute(nums) == permutations(nums) for nums in numbers)
    True
    """
```

Comment (cclauss) [reply data_structures/arrays/permutations.py:2]: @dhruvmanila In the `keeper` can we add a hotlink so that.

> please provide doctest for the function permute

becomes

> please provide [doctest](https://docs.python.org/3/library/doctest.html) for the function permute

Comment (dhruvmanila) [reply data_structures/arrays/permutations.py:2]: They're provided, just not in the review comments. I guess we could move them.

File: data_structures/arrays/permutations.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    numbers = [1, 2, 3]
+    print("permutations are:", permute(numbers))
+    print(
+        "isValid:",
+        all(list(nums) in permute(numbers) for nums in permutations(numbers)),
+    )
+
+
+if __name__ == "__main__":
+    main()
```
Comment (ChrisO345) [inline data_structures/arrays/permutations.py:35]:
```suggestion
if __name__ == "__main__":
    numbers = [1, 2, 3]
    print("permutations are:", permute(numbers))
    print(
        "isValid:",
        all(list(nums) in permute(numbers) for nums in permutations(numbers)),
    )
```

**Review by cclauss — APPROVED** [review]

**Comment by rajansh87** [comment]
Please merge this pr as all checks have passed.

**Review by CaedenPH — APPROVED** [review]

File: data_structures/arrays/permutations.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from itertools import permutations
```
Comment (dhruvmanila) [inline data_structures/arrays/permutations.py:1]: If this is not being used in the module, please remove it. I see it's just used in the `if __name__ == ...` block, add it there instead.

File: data_structures/arrays/permutations.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+
+# All permutations:
+def permute(nums: list[int]) -> list[list[int]]:
+    """
+    >>> from itertools import permutations
+    >>> numbers= [1,2,3]
+    >>> all(list(nums) in permute(numbers) for nums in permutations(numbers))
```
Comment (dhruvmanila) [inline data_structures/arrays/permutations.py:9]: Please add more tests like when the input is an empty list, list with 1 number, multiple numbers, etc. Test out the algorithm as a [black box](https://en.wikipedia.org/wiki/Black-box_testing) and [white box](https://en.wikipedia.org/wiki/White-box_testing).

File: data_structures/arrays/permutations.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    True
+    """
+    result = []
+    if len(nums) == 1:
+        return [nums.copy()]
+    for _ in range(len(nums)):
+        n = nums.pop(0)
```
Comment (dhruvmanila) [inline data_structures/arrays/permutations.py:16]:
```suggestion
    for n in nums:
```

I think you just want to loop over each element in the array?

File: data_structures/arrays/permutations.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        permutations = permute(nums)
+        for perm in permutations:
+            perm.append(n)
+        result.extend(permutations)
+        nums.append(n)
```
Comment (dhruvmanila) [inline data_structures/arrays/permutations.py:21]: You're mutating the input in-place! That is a very bad idea. The function is returning the result, so why are you mutating the input?

**Review by dhruvmanila — DISMISSED** [review]

**Review by ChrisO345 — APPROVED** [review]

**Comment by CaedenPH** [comment]
@rajansh87 Once you commit the last changed that dhruv suggested, your pull request will be ready for merge
